content/blog/2022-05-18-learned-federated-query-optimization.md

```diff
@@ -0,0 +1,178 @@
+and then, it would execute the query its execution engine. However, more complex strategies could be followed for a query 
+like this. For example, `A ⋈ B ⋈ C` could be split into two complement subqueries, `A` (scan) and `B ⋈ C` (join), where 
+the first query which is just a scan would be executed as a simple scan from DS1. However, the second query (`B ⋈ C`) 
+could be pushed-down for local execution in `DS2`. But how do we know if pushing-down `B ⋈ C` to `DS2` would result to
+a better performance compared to just fetching tables `B` and `C` to the federation engine (Spark SQL) and executing the 
+join there?
+
+## Learned Query Compilation
+To answer the previous question, we use a machine learning model as our cost model that helps us evaluate the cost of 
+a query without executing it. Using this model, we can easily evaluate whether it is better to push-down a query or not.
+In order to do that, we use a unified vector space model that is able to transform SQL queries into vectors, and thus,
+we can feed these queries to our model in order to train and predict the execution time of future queries. Thus, our
+optimizer is able to leverage this machine learning model in order to decide which parts of the query should push-down 
+or not.
+
+## Query Compiler Design
+Let's take a look at how our cross-database query compiler works. Similarly to any other compiler, we have a syntax parser
+which takes as input an SQL query in raw text and produces and intermediate representation. Then, we perform further 
+optimizations to this intermediate representation and then we pass the output optimized plan to the executor in order to 
+evaluate the query and get the resulting rows.
+
+### Compilation Lifecycle
+The steps of compilation of a query consists of the following steps.
+1. Parsing: We use Spark SQL query parser
+2. Intermediate Representation: The first intermediate representation is a tree-like structure, where each parent node represents the join between two tables
+3. Graph Representation: We tranform the aforementioned intermediate representation to a more modular graph structure, which gives us more freedom on create efficient plans.
+4. Optimization: By examining the graph, we produce a the optimal (according to the cost model) tree-like query execution plan
+5. Execution: We pass the final plan back to the federation engine for evaluation
+
+We explain in more detail the intermediate representation that our compiler uses.
+
+### Graph Intermediate Representation
+Let's take again a look at the previous example
+```sql
+SELECT AVG(A.salary)
+FROM A, B, C
+WHERE A.id = B.id
+AND A.id = C.id
+AND B.id = C.id
+```
+
+If we think of this query as a graph, it would look as follows:
+```
+    A - B
+    \   / 
+      C
+```
+
+This is a fully-connected graph, as there are connections from all nodes to all nodes. As the final plan will be a binary tree,
+that means that the final plan could be any of the following three trees:
```

Comment:
Again, it would be nice to explicitly say that these trees represent the data flow between tables through operations (joins).